system-helper/flatpak-system-helper.c

```diff
@@ -306,18 +320,15 @@ take_ongoing_pull_by_dir (const gchar *src_dir)
 {
   OngoingPull *pull = NULL;
 
-  G_LOCK (ongoing_pulls);
-  for (guint i = 0; i < ongoing_pulls->len; i++)
-    {
-      OngoingPull *pull_tmp = g_ptr_array_index (ongoing_pulls, i);
-      if (g_strcmp0 (pull_tmp->src_dir, src_dir) == 0)
-        {
-          g_ptr_array_remove (ongoing_pulls, pull_tmp);
-          pull = g_steal_pointer (&pull_tmp);
-          break;
-        }
-    }
-  G_UNLOCK (ongoing_pulls);
+  G_LOCK (cache_dirs_in_use);
+  /* Keep src_dir key inside hashtable but remove its OngoingPull
+   * value and set it to NULL. This way src_dir is still marked
+   * as in-use (as Deploy or CancelPull might be executing on it,
+   * whereas OngoingPull ownership is transferred to respective
+   * callers. */
+  pull = g_hash_table_lookup (cache_dirs_in_use, src_dir);
+  g_hash_table_insert (cache_dirs_in_use, g_strdup (src_dir), NULL);
```

Comment:
This does the wrong thing if `src_dir` is not currently a key of `cache_dirs_in_use`. You could do something like this:

```c
gpointer key, value;
if (g_hash_table_steal_extended (cache_dirs_in_use, src_dir, &key, &value))
  {
    g_hash_table_insert (cache_dirs_in_use, key, NULL);
    pull = value;
  }
```

which also avoids having to copy the key.

Reply:
Review accepted and addressed in the latest round of iteration.